src/inference/ns.jl

```diff
@@ -0,0 +1,354 @@
+## 1. A subtype of AbstractSampler, defined as a mutable struct containing state information or sampler parameters... Partially done in  mutable struct NSState ? check
+## 2. A function sample_init! which performs any necessary set-up (default: do not perform any set-up)... Partially done in: function AbstractMCMC.sample_init! ? check
+## 3. A function step! which returns a transition that represents a single draw from the sampler... Partially done with: 2 function AbstractMCMC.step! ? check
+## 4. A function transitions_init which returns a container for the transitions obtained from the sampler (default: return a Vector{T} of length N where T is the type of the transition obtained in the first step and N is the number of requested samples).
+## 5. A function transitions_save! which saves transitions to the container (default: save the transition of iteration i at position i in the vector of transitions)
+## 6. A function sample_end! which handles any sampler wrap-up (default: do not perform any wrap-up)... Partially done in: function AbstractMCMC.sample_end! ? check
+## 7. A function bundle_samples which accepts the container of transitions and returns a collection of samples (default: return the vector of transitions)... Partially done with: 2 function AbstractMCMC.bundle_samples ? check
+
+## model:: NestedModel
+## sampler:: Nested
+
+## How the struct NestedModel is defined:
+##struct NestedModel <: AbstractModel
+##    loglike::Function
+##    prior_transform::Function
+##end
+
+##function NestedModel(loglike, priors::AbstractVector{<:UnivariateDistribution})
+##    prior_transform(X) = quantile.(priors, X)
+##    return NestedModel(loglike, prior_transform)
+##end
+
+## revisit the Nested sampler definition here: https://github.com/TuringLang/NestedSamplers.jl/blob/master/src/staticsampler.jl
+
+## comment on each step of the code and also include docstrings for sections of the code
+## both NestedModel and Nested are to be utilized in this code
+
+struct NS{space, T, P <: AbstractProposal, B <: AbstractBoundingSpace{T}} <: InferenceAlgorithm  ## refer the comment on line 39
+    ndims::Int    # number of parameters    ## check this step
+    nactive::Int    # number of active points    ## check this step
+    active_us::Matrix{T}
+    active_points::Matrix{T}
+    active_logl::Vector{T}
+    log_vol::Float64
+    logz::Float64
+    h::Float64
+    update_interval::Int
+    since_update::Int
+    enlarge::Float64
+    has_bounds::Bool
+    ncall::Int
+    min_ncall::Int
+    min_eff::Floaat64
+    proposal::P
+    active_bound::B   
+end  
+
+proposal(p::NestedSamplers.Proposals) = p
+bound(b::NestedSamplers.Bounds) = b
+
+NS() = NS{()}()
+NS(space::Symbol) = NS{(space,)}()
+
+isgibbscomponent(::NS) = true # this states that NS alg is allowed as a Gibbs component
+
+## Define a general `funtion NS` here ?
+
+mutable struct NSState{V<:VarInfo} <: AbstractSamplerState   ## where is this being used? in the function Sampler ?
```

Comment:
To answer your comment: The sampler state is where you store state information for the sampler, whatever that could mean. In this case it looks like you only need `vi`, which is provided by the default sampler state here: https://github.com/TuringLang/Turing.jl/blob/614f593424f54f458f992ebe81454c29215f28fc/src/inference/Inference.jl#L528

You can remove this `NSState` stuff and just use `SamplerState`.

Reply:
Scratch that, you should probably put the `Nested` struct in `NSState`, since `Nested` is mutable and seems to be modified pretty heavily.